Domains: Plans only test

umurkontaci opened the pull request:
This PR adds the "Domains with plans only A/B test" on both NUX and Domain Management -> Domain Search.

### Screenshots
![image](https://cloud.githubusercontent.com/assets/991022/14803379/d56f077c-0b0d-11e6-9b8a-2d7efc1af14e.png)

![screen shot 2016-04-25 at 5 54 35 pm](https://cloud.githubusercontent.com/assets/991022/14803498/de7c2600-0b0e-11e6-9529-bd00c33e5d3e.png)

### Enabling the test
`localStorage.setItem('ABTests','{"domainsWithPlansOnly_20000101":"plansOnly"}')`

### Todo:
- [x] CSS Alignment Fixes ("Included in Premium Plan" text) in Domain Suggestions ( @breezyskies )
- [ ] Test distribution & start dates ( @lucasartoni )

### Flows to test:
- NUX flow (/start) (Registration and Mapping)
- Domain Management -> Domain Search (Registration and Mapping)

### Notes:
- You should not see any price details for a domain / mapping, only Premium plan information.
- Any time you add a domain registration / mapping to your cart, Premium plan should be automatically added.
- Any time you remove Premium Plan from your cart, domain registrations / mappings should be removed as well.
- The test should not change the behaviour when it is not active.
- The test should not change the behaviour if you already have a plan.

### Tests & Checks (mark them as you test, anyone can test)
#### NUX
- [x] Did not see any individual domain / mapping price
- [x] Premium Plan was auto added when adding a domain registration
- [x] Premium Plan was auto added when adding a domain mapping
- [x] The flow did not change when the test was disabled
- [x] Removing the Premium Plan from the cart removed the domain product
- [x] Hovering / clicking on the price information text (Included in Premium) showed plan information
- [x] Design looks good on mobile
- [x] Nothing seems broken

#### Domain Management -> Domain Search
- [x] Did not see any individual domain / mapping price
- [x] Premium Plan was auto added when adding a domain registration
- [x] Premium Plan was auto added when adding a domain mapping
- [x] The flow did not change when the test is disabled
- [x] Removing the Premium Plan from the cart removed the domain product
- [x] Hovering / clicking on the price information text (Included in Premium) showed plan information
- [x] In a site with a paid plan, adding a domain did not Premium plan to cart.
- [x] In a site with a paid plan, with the domain credit unused, the domain price was displayed correctly.
- [x] In a site with a paid plan, with the domain credit *used*, the domain price was displayed correctly.
- [x] Design looks good on mobile
- [x] Nothing seems broken

### Design OK
- [x] @breezyskies 

/cc: @breezyskies @p3ob7o @klimeryk @wensco @lucasartoni @aidvu @shaunandrews

p3ob7o commented:
In the Premium details that we display when hovering on the "Included in the Premium Plan" notice, we show the yearly price:
<img width="323" alt="screen shot 2016-04-26 at 10 10 45" src="https://cloud.githubusercontent.com/assets/1436372/14811378/22575c94-0b98-11e6-8210-669632db6651.png">

We have recently shifted to showing monthly rates to all users, I believe that for consistency we shoud do the same here:
<img width="230" alt="screen shot 2016-04-26 at 10 14 56" src="https://cloud.githubusercontent.com/assets/1436372/14811398/3c3015ca-0b98-11e6-8099-0e2b0ee280e1.png">

p3ob7o commented:
The small lock next to tho "Included in the Premium Plan" notice breaks to the next line:
<img width="495" alt="screen shot 2016-04-26 at 10 19 09" src="https://cloud.githubusercontent.com/assets/1436372/14811470/8d2f4914-0b98-11e6-901c-c306132819ae.png">
(Chrome 51.0.2704.22 beta)

The best way to fix this is to remove the lock itself, I don't really understand what it means, and it makes the page visually stressful :)

p3ob7o commented:
Please change "Included in the Premium Plan" into "Included in WordPress.com Premium"

Three reasons to this request:
- Consistency between locales: only "Included" will need translation.
- Push WordPress.com Premium as a product name instead of something generic.
- Avoid the odd "Included in the Business Plan" that will sound strange.

klimeryk commented:
When I try to "Upgrade" to a mapped domain during the NUX flow, I get this exception:
```
Uncaught TypeError: Cannot read property 'products' of undefined
```
in `client/lib/cart-values/cart-items.js:153`

klimeryk commented on client/components/domains/map-domain-step/index.jsx:
```diff
@@ -173,13 +177,20 @@ var MapDomainStep = React.createClass( {
 		}
 	},
 
+	addPremiumPlanToCart() {
```
This function seems [duplicated](https://github.com/Automattic/wp-calypso/pull/5000/files#diff-99f1f1d8ba147cd06cfd4a0d8d195d34R456).

klimeryk commented on client/components/domains/map-domain-step/index.jsx:
```diff
@@ -173,13 +177,20 @@ var MapDomainStep = React.createClass( {
 	},
 
 	addMappingToCart: function( domain ) {
+		this.addPremiumPlanToCart();
```
This needs to wrapped in a conditional on `this.props.withPlansOnly`, otherwise we'll always be adding the Premium plan, even for the `original` variation :D

klimeryk commented:
I've also noticed that you are passing down the component tree the `withPlansOnly` property in a few cases, instead of just using `abtest` in the component that actually needs this property. While it's definitely the "React way", I think in case of A/B tests (where the property is introduced only temporarily into code and should not change during normal user interaction), it's cleaner (less changes) to just query the `abtest` variation where needed. It's also easier to revert the test (or merge the successful variation) when there are less changes (especially to `propTypes`, etc.). Just my two cents :) This is a big change and I don't you to do some extra work if it can be avoided.

klimeryk commented:
@p3ob7o:
> Consistency between locales: only "Included" will need translation.

I don't think that's true, you are not taking into account that some languages have more complex grammar than English :wink: For example, in Polish, it would make it very hard to translate this properly if you just had "Include in $translation_for_wordpress.com_premium" to work with. Not sure what's the correct name for it is in English (declension, I think), but `$translation_for_wordpress.com_premium` would have to be in different form given this context.
I think we should just keep it as a full string and let the translators figure out the best/correct way to translate it.

breezyskies commented:
Added a commit to fix the wrapping issue and do a few design tweaks:

<img width="415" alt="screen shot 2016-04-26 at 7 15 26 am" src="https://cloud.githubusercontent.com/assets/3011211/14821221/c4094d78-0b7e-11e6-947b-7d269cdf1094.png">

I included Paolo's suggested copy change here, which I think is a good for consistency.

mtias commented:
Could we align this work with the PlanCompare cards work done here:

![image](https://cloud.githubusercontent.com/assets/548849/14821613/ba85adda-0bcb-11e6-936a-d693f4d78add.png)

At least stylistically to start with, seems it's tackling a very similar problem. cc @adambbecker @breezyskies

breezyskies commented:
Design updated to be more consistent with the PlanCompare card. If this test does well, we can probably repurpose that component here.

<img width="416" alt="screen shot 2016-04-26 at 8 49 11 am" src="https://cloud.githubusercontent.com/assets/3011211/14824726/dca7a94e-0b8b-11e6-86d4-7e95ea221365.png">

umurkontaci commented:
Everything above is addressed / fixed. I changed the remaining copies to include WordPress.com Premium, used monthly pricing display.

@klimeryk Cleaned up the code, removed `withPlansOnly` prop and fixed the bugs

@breezyskies The alignment issue is still there when we display the price (it moved to the left): 

<img width="770" alt="screen shot 2016-04-26 at 11 10 31 am" src="https://cloud.githubusercontent.com/assets/991022/14832501/7bf2337a-0baf-11e6-85d9-01aa3a0ed7b3.png">

New look:
![image](https://cloud.githubusercontent.com/assets/991022/14832551/b6a6e510-0baf-11e6-86e2-c95b0df7b429.png)

umurkontaci commented:
@breezyskies Per 'New look' do you think we should increase the size of the header or change it to "WordPress.com Premium"?

breezyskies commented:
Fixed the CSS alignment issue and updated the pricing CSS. 

Design 👍

klimeryk commented:
Thanks, Umur - indeed all my previous findings have been fixed. You even removed some dead code and found an old bug too (missing `break`). Kudos! :bow:

Unfortunately, I found a new issue: it seems the case where the user is already on a non-free plan is not properly handled. It was probably like that there from the beginning, but unfortunately I haven't tested this case before :disappointed: 

If you visit the `Domains` page, it's a mix of the two variations (context: site with Premium plan *and* I've already used the bundled credits to buy a domain):
![2016-04-27-130432_1478x673_scrot](https://cloud.githubusercontent.com/assets/3392497/14850836/1f559fbc-0c7c-11e6-88c0-fa677db69bac.png)

Additionally, when you click `Add`, the Premium plan gets added to the cart (despite the fact that the user already has it), so essentially it means every domain costs $99 :wink: You can't remove the plan without removing the domain too.

Similar case when the site is already on Business plan (Premium plan is also getting added to cart then).

I've also noticed two potential issues in signup with this test active:
* after selecting the free `*.wordpress.com` domain, the plan selection page is skipped and your site is immediately getting set up with the Free plan. I'm not sure if this is intentional or a regression. One could argue that the NUX flow thanks to it is smoother, since _maybe_ if the user wants a free domain then we can assume that he/she wants a free plan now, but it will be harder to convert them to Premium later on (?).
* when the user selects a custom domain, the same thing happens - but, of course, the user is set up with a Premium plan. What if the user wanted to get Business? Of course, he/she can upgrade after signup, but this might be confusing for some users.

breezyskies commented:
> after selecting the free *.wordpress.com domain, the plan selection page is skipped and your site is immediately getting set up with the Free plan. I'm not sure if this is intentional or a regression. One could argue that the NUX flow thanks to it is smoother, since maybe if the user wants a free domain then we can assume that he/she wants a free plan now, but it will be harder to convert them to Premium later on (?).

Good catch. I think we should show the plans step still if the user does not select a custom domain, since selecting the *.wordpress.com is not necessarily saying no to a plan.

> when the user selects a custom domain, the same thing happens - but, of course, the user is set up with a Premium plan. What if the user wanted to get Business? Of course, he/she can upgrade after signup, but this might be confusing for some users.

We intentionally simplified this test by focusing on Premium instead of Business, so if they select a custom domain, it's okay to skip the plan selection step.

umurkontaci commented:
Thanks for the review Igor and Brie, I've added the plans step when the user selects a free domain. and fixed the bugs you've mentioned. Can you give it another go?

klimeryk commented:
Tried to break it in every possible (sane and slightly insane) way - LGTM!

mtias commented:
Don't forget to give a round of squashing with these when possible!
